### DISCUSSION TITLE
How to get users' permission in React?

### DISCUSSION THREAD

**Discussion by cesarsmaniottospotlar** [fief-dev/fief#378] [Q&A]
Hello guys

I'm using Fief to authenticate users in my React app and I would like using permissions to control which menus must be available accodingly with user permissions.

But, the return of hook `useFiefUserinfo` doesn't return user permissions

Are there any way to get this information?

Thanks!

**Comment by fief-bailiff** [comment]
Hail, @cesarsmaniottospotlar 👋 Welcome to Fief's kingdom!

Our team will get back to you very soon to help.

In the meantime, take a minute to [star our repository](https://github.com/fief-dev/fief) ⭐️

<p align="center">
<img width="656" alt="star-fief" src="https://github.com/fief-dev/fief/assets/1144727/b2db7f59-102f-47e5-9751-67fa10372c52">
</p>

## Want to support us?

Subscribe to one of our paid plan to help us continue our work and receive exclusive information and benefits! Starts at $5/month 🪙

<p align="center">
    <a href="https://polar.sh/fief-dev">
    <picture>
      <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/embed/subscribe.svg?org=fief-dev&darkmode=1">
      <img alt="Subscribe" src="https://polar.sh/embed/subscribe.svg?org=fief-dev">
    </picture>
    </a>
</p>

Farewell!

**Comment by frankie567** [comment]
You can use [`useFiefAccessTokenInfo`](https://fief-dev.github.io/fief-js/functions/nextjs_react.useFiefAccessTokenInfo.html): it'll return you a [`FiefAccessTokenInfo`](https://fief-dev.github.io/fief-js/interfaces/index.FiefAccessTokenInfo.html) object which contains an array with the permissions:

```json
{
    "id": "aeeb8bfa-e8f4-4724-9427-c3d5af66190e",
    "scope": ["openid", "required_scope"],
    "acr": "1",
    "permissions": ["castles:read", "castles:create", "castles:update", "castles:delete"],
    "access_token": "ACCESS_TOKEN",
}
```

**Reply by NataliaFCavicchioli** [reply]
@frankie567 I am having the same issue. Are you sure this hook still works? Fief does not recognize it.

`'"@fief/fief/react"' has no exported member named 'useFiefAcceskenInfo'.`

**Reply by frankie567** [reply]
It looks like you have a typo: `'useFiefAcceskenInfo'.` The correct name is `useFiefAccessTokenInfo`.

**Reply by NataliaFCavicchioli** [reply]
Sorry about that, but still, doesn't work. 

`'"@fief/fief/react"' has no exported member named 'useFiefAccessTokenInfo'.`

**Reply by frankie567** [reply]
Right, sorry, actually, I confused with the Next.js hooks 😊 It looks like there is a hole here where the React integration doesn't have a hook decoding the access token to read the permission. Tracking it here: https://github.com/fief-dev/fief-js/issues/11

**Reply by NataliaFCavicchioli** [reply]
No worries. Ok, thank you very much for looking into it!

**Reply by ysnbyzli** [reply]
> Right, sorry, actually, I confused with the Next.js hooks 😊 It looks like there is a hole here where the React integration doesn't have a hook decoding the access token to read the permission. Tracking it here: [fief-dev/fief-js#11](https://github.com/fief-dev/fief-js/issues/11)

Hello 👋🏻,

I’ve been working on a similar task and tried to retrieve user permissions in my React app. I want to use Fief for authentication and control the visibility of menus based on user permissions.

I made some attempts using the useFiefUserinfo hook, but it doesn’t return user permissions as expected. Currently, I’m trying to access the user permissions by using the validateAccessToken function and the useFiefAccessTokenInfo hook like this:

```js

browser.ts > FiefAuth

public validateAccessToken(): Promise<FiefAccessTokenInfo> {
    const token = this.getTokenInfo();

    if (!token) {
      throw new FiefAuthNotAuthenticatedError();
    }

    return this.client.validateAccessToken(token?.access_token);
}
```

```js
hooks

export const useFiefAccessTokenInfo = (): [
  FiefAccessTokenInfo | null,
  boolean
] => {
  const { auth } = useContext(FiefAuthContext);
  const [accessTokenInfo, setAccessTokenInfo] =
    useState<FiefAccessTokenInfo | null>(null);
  const [isLoading, setIsLoading] = useState(true);

  useEffect(() => {
    (async () => {
      const response = await auth.validateAccessToken();
      setAccessTokenInfo(response);
      setIsLoading(false);
    })();
  }, [auth]);

  return [accessTokenInfo, isLoading];
};
```

However, I haven’t been able to fully access the user permissions with this solution. Could you let me know if this approach is correct or if there’s another way I should be retrieving the permissions? Any guidance or examples would be really helpful!

Thanks in advance!
